torch/lib/c10d/reducer.cpp

```diff
@@ -1175,6 +1180,11 @@ void Reducer::finalize_bucket_dense(Bucket& bucket) {
   }
 }
 
+void Reducer::save_thread_local_state() {
+  std::lock_guard<std::mutex> guard(mutex_);
+  thread_local_state_ = at::ThreadLocalState();
```

Comment:
should set keep_grad_mode to False?

Reply:
Great catch, thanks! The docs in `ThreadLocalState.h` also clarify this:
```
//  (e.g. not preserved when passing from forward pass into
  //   the autograd engine, autograd engine takes care of grad mode)
```